File: Data_Structures_And_Algorithms-UC_SanDiego/1-Algorithmic_Toolbox/Week5/Programming_Assignments/3_edit_distance/edit_distance.py
# Uses python3

def minElement(el1, el2, el3):
    minEl = -float("inf")
    if (el1<=el2 and el1<=el3):
        minEl = el1
    if (el2<=el1 and el2<=el3):
        minEl = el2
    if (el3<=el1 and el3<=el2):
        minEl = el3
    return minEl

def edit_distance(s, t):
    element = float("inf")
    distance = 0
    tmpVec = list()
    D = list(list())
    for i in range(len(t)+1):
        tmpVec.append(element)
    for i in range(len(s)+1):
        D.append(tmpVec)
    for i in range(len(s) + 1):
        D[i][0] = i
    for i in range(len(t) + 1):
        D[0][i] = i
    for i in range(1, len(s) + 1):
        for j in range(1, len(t) + 1):
            if s[i - 1] == t[j - 1]:
                distance = 0
            else:
                distance = 1
            D[i][j] = minElement(D[i - 1][j] + 1, D[i][j - 1] + 1, D[i - 1][j - 1] + distance)
    return D[len(s)][len(t)]

if __name__ == "__main__":
    print(edit_distance(input(), input()))

File: Data_Structures_And_Algorithms-UC_SanDiego/1-Algorithmic_Toolbox/Week5/Programming_Assignments/3_edit_distance/edit_distance.cpp
#include <iostream>
#include <string>
#include <vector>
#include <limits>

int minNumber(int el1, int el2, int el3){
  int minEl = std::numeric_limits<int>::min();
  if (el1<=el2 && el1<=el3){
    minEl = el1;
  }
  if (el2<=el1 && el2<=el3){
    minEl = el2;
  }
  if (el3<=el1 && el3<=el2){
    minEl = el3;
  }
  return minEl;
}

int edit_distance(const std::string &str1, const std::string &str2) {
  int element = std::numeric_limits<int>::min();
  int distance = 0;
  std::vector<int> tmpVec {};
  std::vector<std::vector<int>> D {};
  for (int i=0; i!=str2.length()+1; i++){
    tmpVec.push_back(element);
  }
  for (int i=0; i!=str1.length()+1; i++){
    D.push_back(tmpVec);
  }
  for (int i=0; i!=str1.length()+1; i++){
    D[i][0] = i;
  }
  for (int i=0; i!=str2.length()+1; i++){
    D[0][i] = i;
  }
  for (int i=1; i!=str1.length()+1; i++){
    for (int j=1; j!=str2.length()+1; j++){
      if (str1[i-1] == str2[j-1]){ distance = 0; }
      else { distance = 1; }
      D[i][j] = minNumber(D[i-1][j]+1, D[i][j-1]+1, D[i-1][j-1]+distance);
    }
  }
  return D[str1.length()][str2.length()];
}

int main() {
  std::string str1;
  std::string str2;
  std::cin >> str1 >> str2;
  std::cout << edit_distance(str1, str2) << std::endl;
  return 0;
}
